interfaces/builtin/custom_device.go

```diff
@@ -52,15 +52,19 @@ var (
 	customDeviceInternalError = errors.New(`custom-device interface internal error`)
 
 	// Validating regexp for filesystem paths
-	customDevicePathRegexp = regexp.MustCompile(`^/[^"@]*$`)
+	customDevicePathRegexp = regexp.MustCompile(`^/[^"]*$`)
+
+	// Regexp to escape unescaped `@` characters in filesystem paths
+	customDeviceAtEscaper = regexp.MustCompile(`([^\\](\\\\)*)@`)
 
 	// Validating regexp for udev device names.
 	// We forbid:
 	// - `|`: it's valid for udev, but more work for us
 	// - `{}`: have a special meaning for AppArmor
+	// - `@`: denotes a variable in AppArmor
```

Comment:
I am not sure this is entirely correct, looking at apparmor.d manpage, a variable is defined to have this format:
```
VARIABLE = '@{' ALPHA [ ( ALPHANUMERIC | '_' ) ... ] '}'
````

I am also told that folks are using `system-files` as a workaround, which means that AppArmor parser may actually be happy with paths containing `@` as long as it's not `@{..}`. I'm guessing, it's likely `@` does not need escaping at all in this case, since the profiles worked for `system-files`? At the same time, I suspect that `system-files` may simply not validate the path correctly, so it'd be great to align both interface implementations to use common code for validation.

Reply:
@alexmurray @jslarraz do you know what are the rules around `@` in profiles is it's not part of a vaiable reference nor denotes an abstract unit socket path?

Reply:
My motivation behind prohibiting `@` in device paths was to preserve the existing behavior, while allowing `@` for filepaths. But there's not a technical reason for this, I just wanted to minimize the change in behavior.

Regarding usage of `@`, I think there are cases where `@` has a special meaning in profiles without being followed by `{`. Some examples from the manpage:

```
       In  apparmor  the address of an abstract unix domain socket begins with the @ character, similar to how they are reported
       (as paths) by netstat -x. The address then follows and may contain pattern matching and any characters including the null
       character. In apparmor null characters must be specified by using an escape sequence \000 or \x00. The  pattern  matching
       is  the  same  as  is  used  by file path matching so * will not match / even though it has no special meaning with in an
       abstract socket name. Eg.

         unix addr=@*,
```

```
         # Allow create and use of abstract and anonymous sockets for profile_name
         unix peer=(label=@{profile_name}),
```

```
         # Allow SOCK_STREAM connect, receive and send on an abstract socket @bar
         # with peer running under profile '/foo'
         unix (connect, receive, send) type=stream peer=(label=/foo,addr="@bar"),

         # Allow accepting connections from and receiving from peer running under
         # profile '/bar' on abstract socket '@foo'
         unix (accept, receive) addr=@foo peer=(label=/bar),
```

But these examples all pertain to unix sockets, not filepaths, so probably it would be fine to not escape `@` in filepaths? Though then we need to ensure that filepaths don't contain `@{`, as right now they can contain `{}` characters (though device paths cannot). I do think it is safe to universally escape `@` symbols, if that would be simpler. I'd be interested to hear what the security people think :)

Reply:
I think abstract sockets have somewhat unusual apparmor names, at least in audit messages. I don't recall how this behaves in the parser though. I could check quickly and report back.